Will Albert  
DSSA5102 Gathering & Warehousing  
Mini Project 1 - Analysis with Pandas and Plotly

### Import Pandas, Numpy and Plot.ly (offline mode)

In [1]:
import pandas as pd
import numpy as np

from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)

### Read data (if file is in same directory as notebook; otherwise, add directory info)
### and have a look at the first five rows

In [2]:
df = pd.read_csv("userssharedsdfschoolimprovement2010grants.csv")
df.head()

School Name      City State  \
0  HOGARTH KINGEEKUK MEMORIAL SCHOOL  SAVOONGA    AK   
1                    AKIACHAK SCHOOL  AKIACHAK    AK   
2                     GAMBELL SCHOOL   GAMBELL    AK   
3               BURCHELL HIGH SCHOOL   WASILLA    AK   
4                       AKIAK SCHOOL     AKIAK    AK   

                               District Name 2010/11/Award Amount  \
0              BERING STRAIT SCHOOL DISTRICT           $471014.00   
1                     YUPIIT SCHOOL DISTRICT           $520579.00   
2              BERING STRAIT SCHOOL DISTRICT           $449592.00   
3  MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT           $641184.00   
4                     YUPIIT SCHOOL DISTRICT           $399686.00   

   Model Selected                                           Location  
0  Transformation  200 MAIN ST\nSAVOONGA, AK 99769\n(63.6687, -17...  
1  Transformation  AKIACHAK 51100\nAKIACHAK, AK 99551\n(60.8911, ...  
2  Transformation  169 MAIN ST\nGAMBELL, AK 99742\n(63.7413, -171...  
3  Transformation  1775 WEST PARKS HWY\nWASILLA, AK 99654\n(61.57...  
4  Transformation     AKIAK 5227\nAKIAK, AK 99552\n(60.8879, -161.2)

### Show basic descriptives for each column/variable

In [3]:
df.describe()

School Name          City State         District Name  \
count                  831           831   831                   831   
unique                 821           424    50                   421   
top     BURBANK ELEMENTARY  PHILADELPHIA    CA  PHILADELPHIA CITY SD   
freq                     2            29    92                    27   

       2010/11/Award Amount  Model Selected  \
count                   757             825   
unique                  651               4   
top                   $2.00  Transformation   
freq                     17             608   

                                                 Location  
count                                                 831  
unique                                                823  
top     6941 NORTH CENTRAL ST\nPORTLAND, OR 97203\n(45...  
freq                                                    3

### Looks like only 4 types of "Model Selected"; let's see what they are

In [4]:
df["Model Selected"].unique()

array(['Transformation', 'Restart', 'Turnaround', 'Closure', nan],
      dtype=object)

### OK, how many of each?

In [5]:
model = df["Model Selected"]
model.value_counts()

Transformation    608
Turnaround        168
Restart            33
Closure            16
Name: Model Selected, dtype: int64

### Now let's look at amounts.  A histogram to show frequency distribution...
- Also, I want to make a new column for "Amount", to convert it to numeric values and to rename the column for easier analysis.

In [6]:
df["amount"] = pd.to_numeric(df["2010/11/Award Amount"].str.replace("$", ""))
amt = df["amount"]
iplot(
    go.Figure(
        data=[
            go.Histogram(x=amt)
        ],
        layout=dict(title="Figure 1: Frequency of Amounts, $50K Increments")
    )
)


### Interesting; a lot of amounts on the low end.  We'll come back to that.
### Now I want to see a breakdown of the types of "Models" by State

In [7]:
state_model = pd.crosstab(index=df["State"], columns=model)
print(state_model)

Model Selected  Closure  Restart  Transformation  Turnaround
State                                                       
AK                    0        1               6           0
AL                    0        0              11           0
AR                    0        0               7           0
AZ                    0        0              12           7
CA                    2        5              56          29
CO                    3        1               9           6
CT                    0        1               7           6
DC                    0        3               3           4
DE                    0        0               2           0
FL                    0        0              54          17
GA                    0        0              24           2
IA                    0        0               6           0
ID                    0        0               6           0
IL                    0        1               5           4
IN                    0 

### Can't think of much to do with that information right now.  
### I think I just want to see some basic statistics for "Amount": 
- number of schools
- minimum amount
- maximum amount
- mean
- median
- standard deviation.

In [8]:
n = amt.count()
high = amt.max()
low = amt.min()
mean = amt.mean()
median = amt.median()
sd = amt.std()

border = "=================================="

print(border)
print("Number of schools =", n)
print("Maximum =", high)
print("Minimum =", low)
print("Mean =", round(mean, 2))
print("Median =", round(median, 2))
print("Standard Deviation =", round(sd, 2))
print(border)
print("Table 1: Basic Summary Statistics for 'Amount'")

Number of schools = 757
Maximum = 997852.0
Minimum = 1.0
Mean = 420827.89
Median = 407794.0
Standard Deviation = 287458.05
Table 1: Basic Summary Statistics for 'Amount'


### About those low values from the histogram... 
- Sort by amount
- Look at the first bunch of cases, then the next, and see if there are any patterns.
- I went by groups of 10, which showed that the first 40 give us some information.


In [16]:
df.sort_values('amount').iloc[0:39]

School Name              City State  \
643               EDISON HS/FAREIRA SKILLS      PHILADELPHIA    PA   
80                       BARTON ELEMENTARY    SAN BERNARDINO    CA   
553  WOODWARD CAREER TECHNICAL HIGH SCHOOL        CINCINNATI    OH   
538               HUGHES ELEMENTARY SCHOOL          SYRACUSE    NY   
279                      NORTH HIGH SCHOOL        DES MOINES    IA   
48                            PACIFIC HIGH    SAN BERNARDINO    CA   
531                  EMERSON MIDDLE SCHOOL           YONKERS    NY   
541                    BENNETT HIGH SCHOOL           BUFFALO    NY   
528                   INTERNATIONAL SCHOOL           BUFFALO    NY   
527             DELAWARE ELEMENTARY SCHOOL          SYRACUSE    NY   
545              GEORGE FOWLER HIGH SCHOOL          SYRACUSE    NY   
524                  ROOSEVELT HIGH SCHOOL           YONKERS    NY   
95                    SHANDIN HILLS MIDDLE    SAN BERNARDINO    CA   
53                       SAN GORGONIO HIGH    SAN BERNARDINO    CA   
502                        CROWNPOINT HIGH        CROWNPOINT    NM   
521                 SOUTH PARK HIGH SCHOOL           BUFFALO    NY   
759                     T.C. WILLIAMS HIGH        ALEXANDRIA    VA   
503                      ERNIE PYLE MIDDLE       ALBUQUERQUE    NM   
500                 EL CAMINO REAL ACADEMY       ALBUQUERQUE    NM   
658                             LINCOLN HS      PHILADELPHIA    PA   
506                           NEWCOMB HIGH           NEWCOMB    NM   
499                      LAGUNA-ACOMA HIGH        NEW LAGUNA    NM   
529       FRANKLIN D ROOSEVELT HIGH SCHOOL          BROOKLYN    NY   
278                    WEEKS MIDDLE SCHOOL        DES MOINES    IA   
536                   FLUSHING HIGH SCHOOL          FLUSHING    NY   
523           LONG ISLAND CITY HIGH SCHOOL  LONG ISLAND CITY    NY   
281                     HOYT MIDDLE SCHOOL        DES MOINES    IA   
115                     ARROYO VALLEY HIGH    SAN BERNARDINO    CA   
629                    MCKEESPORT AREA SHS       MC KEESPORT    PA   
436              W A HIGGINS MIDDLE SCHOOL        CLARKSDALE    MS   
505                    RAMIREZ THOMAS ELEM          SANTA FE    NM   
433             LEFLORE COUNTY HIGH SCHOOL         ITTA BENA    MS   
438                PORT GIBSON HIGH SCHOOL       PORT GIBSON    MS   
577                    WELCOME CENTER HIGH          COLUMBUS    OH   
791                GRANDVIEW MIDDLE SCHOOL         GRANDVIEW    WA   
67                  PHARIS F. FEDDE MIDDLE  HAWAIIAN GARDENS    CA   
812                BRADLEY TECHNOLOGY HIGH         MILWAUKEE    WI   
569                SOUTHMOOR MIDDLE SCHOOL          COLUMBUS    OH   
556              KEIFER ALTERNATIVE CENTER       SPRINGFIELD    OH   

                                   District Name 2010/11/Award Amount  \
643                         PHILADELPHIA CITY SD                $1.00   
80                   SAN BERNARDINO CITY UNIFIED                $2.00   
553                              CINCINNATI CITY                $2.00   
538                SYRACUSE CITY SCHOOL DISTRICT                $2.00   
279  DES MOINES INDEPENDENT COMM SCHOOL DISTRICT                $2.00   
48                   SAN BERNARDINO CITY UNIFIED                $2.00   
531                 YONKERS CITY SCHOOL DISTRICT                $2.00   
541                 BUFFALO CITY SCHOOL DISTRICT                $2.00   
528                 BUFFALO CITY SCHOOL DISTRICT                $2.00   
527                SYRACUSE CITY SCHOOL DISTRICT                $2.00   
545                SYRACUSE CITY SCHOOL DISTRICT                $2.00   
524                 YONKERS CITY SCHOOL DISTRICT                $2.00   
95                   SAN BERNARDINO CITY UNIFIED                $2.00   
53                   SAN BERNARDINO CITY UNIFIED                $2.00   
502                  GALLUP-MCKINLEY CTY SCHOOLS                $2.00   
521                 BUFFALO CITY SCHOOL DISTRICT              

### Interesting that there are a bunch of 1's and 2's, then some amounts in the teens, then 100's, then 1000's...
- I'm thinking this is an issue of scale; the 1's and 2's should have been x100000, the teens x10000, and so on...
- Without more information, though, I don't feel like I can make any corrections.
#### Actually, if I look back at my histogram, it looks like if I remove amounts < 200K, it may be close to a normal distribution (though a bit flat).  Still, I don't know enough about the collection methods at this point to feel comfortable doing that.

### We know from our initial description of the file that we have some missing amounts (831 entries vs. 757 amounts, so we should find 74 missing).  Let's look at those.

In [10]:
no_amt = pd.DataFrame(df[df['amount'].isnull()])
no_amt.head(10)

School Name           City State  \
280             LINCOLN ELEMENTARY SCHOOL       WATERLOO    IA   
283             EDMUNDS FINE ARTS ACADEMY     DES MOINES    IA   
285           FORT HALL ELEMENTARY SCHOOL      POCATELLO    ID   
286                   MELBA MIDDLE SCHOOL          MELBA    ID   
287            LAKESIDE ELEMENTARY SCHOOL         WORLEY    ID   
288            SNAKE RIVER JR HIGH SCHOOL      BLACKFOOT    ID   
323  SARAH TOWLES REED SENIOR HIGH SCHOOL    NEW ORLEANS    LA   
324                ASSUMPTION HIGH SCHOOL  NAPOLEONVILLE    LA   
348     GOVERNOR JAMES B LONGLEY ELEM SCH       LEWISTON    ME   
349               LAKE REGION HIGH SCHOOL         NAPLES    ME   

                                   District Name 2010/11/Award Amount  \
280                WATERLOO COMM SCHOOL DISTRICT                  NaN   
283  DES MOINES INDEPENDENT COMM SCHOOL DISTRICT                  NaN   
285                           BLACKFOOT DISTRICT                  NaN   
286                         MELBA JOINT DISTRICT                  NaN   
287                PLUMMER-WORLEY JOINT DISTRICT                  NaN   
288                         SNAKE RIVER DISTRICT                  NaN   
323                 RECOVERY SCHOOL DISTRICT-LDE                  NaN   
324                            ASSUMPTION PARISH                  NaN   
348                   LEWISTON SCHOOL DEPARTMENT                  NaN   
349                                      MSAD 61                  NaN   

     Model Selected                                           Location  amount  
280  Transformation  302 CEDAR BEND\nWATERLOO, IA 50703\n(42.5186, ...     NaN  
283  Transformation  1601 CROCKER ST\nDES MOINES, IA 50314\n(41.594...     NaN  
285  Transformation      B ST\nPOCATELLO, ID 83202\n(42.906, -112.582)     NaN  
286  Transformation   560 BRDWAY\nMELBA, ID 83641\n(43.3785, -116.531)     NaN  
287  Transformation  S 29900 FIRST AVE\nWORLEY, ID 83876\n(47.4, -1...     NaN  
288  Transformation  918 WEST HWY 39\nBLACKFOOT, ID 83221\n(43.197,...     NaN  
323  Transformation  5316 MICHOUD BLVD\nNEW ORLEANS, LA 70129\n(30....     NaN  
324  Transformation  4880 HWY 308\nNAPOLEONVILLE, LA 70390\n(29.941...     NaN  
348      Turnaround  145 BIRCH ST\nLEWISTON, ME 4240\n(44.0943, -70...     NaN  
349  Transformation  1877 ROOSEVELT TRAIL\nNAPLES, ME 4055\n(43.996...     NaN

### There they are.  Nothing sticks out at first glance, so let's "describe" the data frame.

In [11]:
no_amt.describe(include='all')

School Name         City State District Name  2010/11/Award Amount  \
count           74           74    74            74                   0.0   
unique          74           47     9            51                   0.0   
top     TRAVIS H S  SAN ANTONIO    TX   HOUSTON ISD                   NaN   
freq             1            8    48             6                   NaN   
mean           NaN          NaN   NaN           NaN                   NaN   
std            NaN          NaN   NaN           NaN                   NaN   
min            NaN          NaN   NaN           NaN                   NaN   
25%            NaN          NaN   NaN           NaN                   NaN   
50%            NaN          NaN   NaN           NaN                   NaN   
75%            NaN          NaN   NaN           NaN                   NaN   
max            NaN          NaN   NaN           NaN                   NaN   

        Model Selected                                           Location  \
count               68                                                 74   
unique               2                                                 74   
top     Transformation  2524 WEST LEDBETTER DR\nDALLAS, TX 75233\n(32....   
freq                65                                                  1   
mean               NaN                                                NaN   
std                NaN                                                NaN   
min                NaN                                                NaN   
25%                NaN                                                NaN   
50%                NaN                                                NaN   
75%                NaN                                                NaN   
max                NaN                                                NaN   

        amount  
count      0.0  
unique     NaN  
top        NaN  
freq       NaN  
mean       NaN  
std        NaN  
min        NaN  
25%        NaN  
50%        NaN  
75%        NaN  
max        NaN

#### So, we knew there were 74 missing amounts, but now we also know that those missing amounts consist of 74 unique schools in 51 unique districts, in 47 cities, across 9 states.

## At this point, I want to look at the breakdown of amounts for the "Model Selected" and "State" variables.

### First, I'll look at the mean amount for each type of model, keeping in mind that the means will be affected by those small amounts.

In [12]:
model_means = df['amount'].groupby(df["Model Selected"]).mean()

print(border)
print(round(model_means, 2))
print(border)
print("Table 2: Mean Amounts by Model Type")

Model Selected
Closure            75155.50
Restart           503412.42
Transformation    434116.06
Turnaround        394100.58
Name: amount, dtype: float64
Table 2: Mean Amounts by Model Type


### Then, I want to see the total grant funds distributed for each state...

In [23]:
state_sums = df['amount'].groupby(df['State']).sum().reset_index()
state_sums

State      amount
0     AK   3607416.0
1     AL   2300782.0
2     AR   3936109.0
3     AZ  13181345.0
4     CA  20286824.0
5     CO  10158232.0
6     CT   8041137.0
7     DC   4199339.0
8     DE   1303216.0
9     FL  43138859.0
10    GA   8892929.0
11    IA    761235.0
12    ID    496680.0
13    IL   2197110.0
14    IN   2191767.0
15    KS   1270660.0
16    KY   5025938.0
17    LA         0.0
18    MA   7441266.0
19    MD   6033577.0
20    ME         0.0
21    MI  10893630.0
22    MN  11633397.0
23    MO  13215147.0
24    MS   1250153.0
25    MT   1961690.0
26    NC  13527138.0
27    ND    418574.0
28    NE   4078079.0
29    NH   2661604.0
30    NJ   2158303.0
31    NM   1975170.0
32    NV   4991568.0
33    NY   5668851.0
34    OH  15266786.0
35    OK   2770485.0
36    OR   5792028.0
37    PA  28318207.0
38    RI         0.0
39    SC  11272707.0
40    SD    530890.0
41    TN   4997708.0
42    TX         0.0
43    UT   4251625.0
44    VA   8008990.0
45    VT   2750483.0
46    WA   6569211.0
47    WI   3176096.0
48    WV   4961093.0
49    WY   1002682.0

#### ... which would look much better on a map.

In [24]:
iplot(
    go.Figure(
        data = [ dict(
            type='choropleth',
            autocolorscale = True,
            locations = state_sums['State'],
            z = state_sums['amount'],
            locationmode = 'USA-states',
            marker = dict(
            line = dict (
                    color = 'rgb(255,255,255)',
                    width = 2
                ) ),
            colorbar = dict(
                title = "USD")
            ) ],

        layout = dict(
            title = 'Figure 2: 2010 School Improvement Grants by State',
            geo = dict(
                scope='usa',
                projection=dict( type='albers usa' ),
                showlakes = True,
                lakecolor = 'rgb(255, 255, 255)'),
             )
    )
)

#### We can clearly see that Florida got the most, followed by Pennsylvania.  A good follow-up to this would be to see how many schools are in each state.

### We could further break this down with the city variable, or use the latitude/longitude data contained in the "Location" variable to plot individual schools.